File: lib/python/opentoken-core-ai/src/main/opentoken_core_ai/tokentransformer/rotation/rotation_embedding_transformer.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from threading import Lock
from typing import List, Optional

from opentoken_core_ai.tokentransformer.rotation.embedding_rotator import rotate
from opentoken_core_ai.tokentransformer.rotation.rotation_matrix_generator import generate
from opentoken_core_ai.tokentransformer.rotation.rotation_quantizer import quantize


class RotationEmbeddingTransformer:
    """Composites matrix generation, rotation projection, and quantization into a
    single float-vector → token-list transformer.

    Rotation matrices are generated lazily from the IV and cached for the lifetime
    of the instance. Thread-safe initialization.
    """

    def __init__(
        self,
        iv: str,
        rotation_count: int,
        dimension: int,
        hash_dimension: int,
        bias: Optional[List[float]] = None,
        min_val: float = -5.0,
        max_val: float = 5.0,
        bin_width: float = 0.05,
    ):
        """Initialize the transformer with rotation configuration.

        Args:
            iv: Initialization vector string used to derive rotation matrices.
            rotation_count: Number of rotation matrices to generate.
            dimension: Size of the input embedding vector (N).
            hash_dimension: Number of projected dimensions to retain and quantize (k ≤ N).
            bias: Optional float vector of length N subtracted before rotation. Defaults to zeros.
            min_val: Quantizer lower bound (default -5.0).
            max_val: Quantizer upper bound (default +5.0).
            bin_width: Quantizer bin width (default 0.05).
        """
        self._iv = iv
        self._rotation_count = rotation_count
        self._dimension = dimension
        self._hash_dimension = hash_dimension
        self._bias = bias if bias is not None else [0.0] * dimension
        self._min_val = min_val
        self._max_val = max_val
        self._bin_width = bin_width
        self._matrices: Optional[List[List[List[float]]]] = None
        self._lock = Lock()

    def transform(self, embedding: List[float]) -> List[str]:
        """Transform a raw float embedding into rotation-quantized token strings.

        Returns one token string per rotation matrix.

        Args:
            embedding: Raw float vector of length N (must match the configured dimension).

        Returns:
            List of space-separated bin-index strings, one per rotation matrix.
        """
        self._ensure_matrices()
        projections = rotate(embedding, self._matrices, self._bias, self._hash_dimension)
        return [quantize(p, self._min_val, self._max_val, self._bin_width) for p in projections]

    def _ensure_matrices(self) -> None:
        """Lazily generate and cache rotation matrices in a thread-safe manner."""
        if self._matrices is None:
            with self._lock:
                if self._matrices is None:
                    self._matrices = generate(self._iv, self._rotation_count, self._dimension)

File: lib/java/openlinktoken-core-ai/src/main/java/org/openlinktoken/tokentransformer/rotation/RotationEmbeddingTransformer.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.tokentransformer.rotation;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Transforms a float embedding into a list of quantized rotation token strings.
 *
 * <p>The transformation pipeline for each configured rotation matrix is:
 * <ol>
 *   <li>Subtract per-dimension {@code bias} from the embedding (centring).</li>
 *   <li>Project the centred vector through the rotation matrix to obtain a
 *       {@code hashDimension}-dimensional float vector.</li>
 *   <li>Quantize the projected vector into a space-separated integer string.</li>
 * </ol>
 *
 * <p>Rotation matrices are generated lazily on the first call to
 * {@link #transform(float[])} and cached for subsequent calls.
 */
public final class RotationEmbeddingTransformer implements EmbeddingTransformer {

    private final String iv;
    private final int rotationCount;
    private final int dimension;
    private final int hashDimension;
    private final float[] bias;
    private final double minVal;
    private final double maxVal;
    private final double binWidth;

    /** Lazily initialized rotation matrices; volatile for double-checked locking. */
    private volatile List<double[][]> matrices;

    /**
     * Constructs a {@code RotationEmbeddingTransformer} with full parameter control.
     *
     * @param iv            initialization vector used to seed the rotation matrices
     * @param rotationCount number of rotation matrices (= number of output tokens)
     * @param dimension     dimension of the rotation matrices (must equal embedding.length)
     * @param hashDimension number of projected dimensions to quantize (k ≤ dimension)
     * @param bias          per-dimension bias subtracted before projection; length must equal dimension
     * @param minVal        quantizer lower bound
     * @param maxVal        quantizer upper bound
     * @param binWidth      quantizer bin width; must be &gt; 0
     * @throws IllegalArgumentException if any parameter is invalid
     */
    public RotationEmbeddingTransformer(
            String iv,
            int rotationCount,
            int dimension,
            int hashDimension,
            float[] bias,
            double minVal,
            double maxVal,
            double binWidth) {
        if (iv == null || iv.isBlank()) {
            throw new IllegalArgumentException("iv must not be null or blank");
        }
        if (rotationCount <= 0) {
            throw new IllegalArgumentException("rotationCount must be > 0");
        }
        if (dimension <= 0) {
            throw new IllegalArgumentException("dimension must be > 0");
        }
        if (hashDimension <= 0 || hashDimension > dimension) {
            throw new IllegalArgumentException("hashDimension must be in (0, dimension]");
        }
        if (bias == null || bias.length != dimension) {
            throw new IllegalArgumentException("bias must be non-null and have length == dimension");
        }
        if (binWidth <= 0) {
            throw new IllegalArgumentException("binWidth must be > 0");
        }

        this.iv = iv;
        this.rotationCount = rotationCount;
        this.dimension = dimension;
        this.hashDimension = hashDimension;
        this.bias = bias.clone();
        this.minVal = minVal;
        this.maxVal = maxVal;
        this.binWidth = binWidth;
        this.matrices = null;
    }

    /**
     * Creates a {@code RotationEmbeddingTransformer} with zero bias and default
     * quantizer parameters (range [-5.0, 5.0), bin width 0.05).
     *
     * @param iv            initialization vector
     * @param rotationCount number of rotation matrices
     * @param dimension     dimension of rotation matrices
     * @param hashDimension number of projected dimensions to quantize
     * @return new transformer instance
     */
    public static RotationEmbeddingTransformer withDefaults(
            String iv, int rotationCount, int dimension, int hashDimension) {
        float[] zeroBias = new float[dimension];
        return new RotationEmbeddingTransformer(
                iv, rotationCount, dimension, hashDimension, zeroBias,
                RotationQuantizer.DEFAULT_MIN, RotationQuantizer.DEFAULT_MAX, RotationQuantizer.DEFAULT_BIN_WIDTH);
    }

    /**
     * Transform an embedding into a list of quantized rotation token strings.
     *
     * <p>Returns {@code rotationCount} tokens, each a space-separated string of
     * {@code hashDimension} integer bin indices.
     *
     * @param embedding raw CLS float embedding vector; length must equal {@code dimension}
     * @return list of quantized token strings, one per rotation matrix
     */
    @Override
    public List<String> transform(float[] embedding) {
        ensureMatrices();
        List<float[]> projections = EmbeddingRotator.rotate(embedding, matrices, bias, hashDimension);
        return projections.stream()
                .map(p -> RotationQuantizer.quantize(p, minVal, maxVal, binWidth))
                .collect(Collectors.toList());
    }

    private synchronized void ensureMatrices() {
        if (matrices == null) {
            matrices = RotationMatrixGenerator.generate(iv, rotationCount, dimension, hashDimension);
        }
    }
}
